### PR TITLE
Ch 7 unit tests pass

### PR DESCRIPTION
**Pull request by oldfartdeveloper**
Issue #109

## TLDR: I have all but three of the exercises covered by unit tests:

1. Exercise Group 1 - Exercise 2: Convince yourself about `lift3`.  This is proven by failed compilation; at some point I'll find a way to write a unit test that can respond to failed compilations.
1. Exercise Group 3 - Exercise 1: The `Traversable` instance for a binary tree.  I want to read more of the book so that I can write the tests for this as @milesfrain and @hdgarrood have suggested.
1. Exercise Group 3 - Exercise 3: `traverse` implemented in `sequence` and vice-versa.  I want to go on and come back to this.

### Initial Test

I needed to add 

```haskell
instance eqPhoneType :: Eq PhoneType where
  eq HomePhone HomePhone = true
  eq WorkPhone WorkPhone = true
  eq CellPhone CellPhone = true
  eq OtherPhone OtherPhone = true
  eq _ _ = false
```
to **AddressBook.purs**  to get the unit tests to run.  Hence, in addition I wrote a unit test to test this as the proof that unit tests are set up for this chapter.

### Changes to `AddressBook.purs` and `Validation.purs`

Brought in @milesfrain simplifications to these two packages.

### Brought in @milesfrain changes in branch `master`

So this PR will replace this branch's `test/chapter7.md` with the `master` branch one.

## Still to be Done

* I need to submit the `master` branch **chapter 7** changes; will hopefully get that done today.
* I'm still not sure how much to modify the exercises in the text to make it easier for the reader to understand that there are unit tests to verify his work.  I know @milesfrain is working on this problem; should probably discuss w/ him.

### CODE REVIEW

**Comment by hdgarrood** [comment]
> I needed to add `instance eqPhoneType :: Eq PhoneType where ...`

Did you know that you can write
```purescript
derive instance eqPhoneType :: Eq PhoneType
```
and have the compiler write this code for you?

**Comment by oldfartdeveloper** [comment]
@hdgarrood Dang, you're right; I forgot all about `derive`.  I'll update, thanks!

**Comment by milesfrain** [comment]
>     1. I'll find a way to write a unit test that can respond to failed compilations
I don't think that's possible, but don't worry about it. We can either just make a note about exercises without tests, remove the exercise, or replace with a new testable exercise.



>     2. The `Traversable` instance for a binary tree.  I want to read more of the book so that I can write the tests for this

I believe all the prerequisites for that are covered in Ch6, but feel free continue on with the book.

>     * I'm still not sure how much to modify the exercises in the text to make it easier for the reader to understand that there are unit tests

A [guide on running tests](https://github.com/milesfrain/purescript-book/blob/ch2-consolidate/text/chapter2.md) will be part of Ch2 (PR #144), so I don't think we need to write much more about in in Ch7.

**Comment by milesfrain** [comment]
I rebased the solutions branch on master to remove the distracting newtype wrapper changes from this PR. You'll want to rebase your solutions on the latest solutions branch too to re-sync everything.

**Comment by milesfrain** [comment]
In order to not confuse readers when adding this to the AddressBook.purs file, I think it would be good to cover `derive` in Ch6 (#149). It may also be good to note that this is only added to facilitate testing.
```purs
derive instance eqPhoneType :: Eq PhoneType
```

**Comment by oldfartdeveloper** [comment]
How about just a comment next to the statement containing the `derive` that this is needed only by the unit tests and the `derive` will be covered in chapter 10?

**Comment by milesfrain** [comment]
Just a comment on `derive` is good for now

**Comment by oldfartdeveloper** [comment]
@milesfrain After your rebase, there were some merge conflicts which I resolved.  Travis passed.

**Comment by oldfartdeveloper** [comment]
Per PR #147, modified this one's first unit test to match the `master` branch.

File: exercises/chapter4/test/Main.purs [hunk 1]
```diff
@@ -6,6 +6,25 @@ import Data.Maybe (Maybe(..))
 import Data.Path (filename, root)
 import Data.Tuple (fst)
 import Effect (Effect)
+import Test.Solutions
+  ( allTrue
```
Comment (milesfrain) [inline exercises/chapter4/test/Main.purs:10]: It doesn't seem like this is rebased correctly on the latest solutions (or master) branch, since there shouldn't be any changes here outside of ch7

**Comment by milesfrain** [comment]
I'll need to rebase solutions on master again. But to spare the hassle in coordinating that, feel free to just redirect this PR to the `master` branch, and I'll take it from there.
I didn't anticipate the added complications of involving multiple contributors in these rebases.

File: exercises/chapter4/test/Main.purs (hunk 1 above)
Comment (oldfartdeveloper) [reply exercises/chapter4/test/Main.purs:10]: Yes, I've noticed (and wondered about) that as well.  It's become tricky.

**Comment by oldfartdeveloper** [comment]
When I switch the target to `master`, GitHub gives me a stern warning that this could be dangerous.  It will probably break the build.  I'll go ahead and switch the target to `master` and pray for your survival. ;)

**Comment by milesfrain** [comment]
@oldfartdeveloper 
I made a bunch of exercise changes to Ch7 in #153 and also merged the solutions.
The remaining exercises should be a lot clearer now, so give them at shot at your convenience.
